Fix merging component templates with a mix of dotted and nested object mapper definitions

felixbarny opened the pull request:
Fixes https://github.com/elastic/elasticsearch/issues/105482

## The background

This regression was introduced in https://github.com/elastic/elasticsearch/pull/97317. That PR added a new way how component templates get merged. When creating mappings for an index that's about to be created, we previously merged component templates one-by-one. The issue with this sequential merging approach is that some settings (like `subobjects: false`) in one component template may affect how other component templates need to be parsed. This makes it dependent on ordering. Consider that there are component templates c1 and c2 that are merged sequentially. In c1, `subobjects` is set to false. This affects how c2 needs to be parsed. However, if c2 sets `subobjects: false`, before the PR, it couldn't affect how c1 is parsed, because it has already been parsed before considering c2.

The PR introduced a way to bulk-merge multiple component templates at once (see the [changes in MetadataCreateIndexService](https://github.com/elastic/elasticsearch/pull/106077#diff-c02a1326ed501a89ce49b7e7a27df4bb9a2256785d00d7278bccb3609334b939)). In the first step, the mapping sources are merged (merging of a `Map<String, Object>`). This ensures that the `subobjects` parameter, for example, is determined for the combination of the mappings, before the mapping source is actually parsed into a `RootObjectMapper`.

So far, all is well and good.

When there are two component templates that have a mix of dotted and object notation, their sources get merged like this:

```json
{
  "properties": {
    "parent.child": {
      "type": "keyword"
    },
    "parent" : {
      "properties" : {
        "child" : {
          "type" : "integer"
        }
      }
    }
  }
}
```

Still, so far, so good.

When parsing the mappings, the `RootObjectMapper` will then have two `ObjectMapper.Builder` instances for the `parent` object. This is also not issue so far, because there's code in `ObjectMapper.Builder#buildMappers` that deals with that exact scenario and merges the two objects:

https://github.com/elastic/elasticsearch/blob/a97552b6f883fccbbb73eb7e4d130d88b9e8465e/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L160-L167

This is where the trouble starts: when looking at the implementation of `ObjectMapper#merge`, it the merge reason is hard-coded to `MAPPING_UPDATE`.

https://github.com/elastic/elasticsearch/blob/a97552b6f883fccbbb73eb7e4d130d88b9e8465e/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L458-L460

But wait, we're still in a context where the merge reason should be `INDEX_TEMPLATE` - we're merging multiple component templates.

Why does the merge reason matter?

Within the merge process, the merge reason determines how to deal with field mappers for the same field:

https://github.com/elastic/elasticsearch/blob/a97552b6f883fccbbb73eb7e4d130d88b9e8465e/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L592-L598

Because the wrong merge reason is selected, the two different definitions of the `child` field (`{"child" : {"type" : "keyword"}}` and `{"child" : {"type" : "integer"}}` ) are merged instead of replaced. However, it's not allowed to change the type of a field when merging two different field mappers. This is why this saga ends with an exception being thrown here:

https://github.com/elastic/elasticsearch/blob/179739effdad01e0f9e35b09799cec07603e05bf/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L404-L408

## The fix

The PR https://github.com/elastic/elasticsearch/pull/97317 made crucial improvements on how component templates get merged. We don't want to roll these back. Instead, the focus of the bug fix is to preserve the right merge reason, depending on the context.

To do so, I've added the merge reason to `MapperBuilderContext`. This seemed like the most natural choice to hand down the merge reason and avoids having to add another `MergeReason` parameter to even more methods. This merge reason is the one to use when merging mappers while building the mapper (when calling `ObjectMapper#merge` in `ObjectMapper.Builder#buildMappers`).

When creating a `MapperBuilderContext`, a `MergeReason` can be supplied. I kept the `public static MapperBuilderContext root(boolean isSourceSynthetic, boolean isDataStream)` method which is using a `MAPPING_UPDATE` merge reason by default. This is mainly so that I don't clutter this PR with a gazillion updated callers in test code. There are also a couple of callers in production code that I didn't feel the need to update by explicitly providing the `MAPPING_UPDATE` merge reason. YMMV, but previously the entire code base was implicitly assuming a `MAPPING_UPDATE` merge reason for `ObjectMappepr#merge` and this is now adding the ability to specify a different one. But in most places, `MAPPING_UPDATE` is the appropriate merge reason.

felixbarny committed:
Fix merging component templates with a mix of dotted and nested object mapper definitions

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @felixbarny, I've created a changelog YAML for you.

felixbarny committed:
Update docs/changelog/106077.yaml

ruflin commented on server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java:
```diff
@@ -1707,6 +1707,78 @@ public void testExpandDottedNotationToObjectMappers() throws IOException {
+        CompressedXContent mapping1 = new CompressedXContent("""
+            {
+              "properties": {
+                "parent.child": {
```
To be safe, could you also add a test for the invers? Have the nested notation first (mapping2) and then the dot notation (mapping1). In theory, should not make a difference but I rather have one test to much then too little.

felixbarny replied:
Sure! I added that scenario with [cb28cdb](https://github.com/elastic/elasticsearch/pull/106077/commits/cb28cdbe5a47b0d3b527087acc192abfa387d266).

felixbarny committed:
Test nested then dotted

felixbarny committed:
Merge remote-tracking branch 'origin/main' into fix-template-merging-regression

eyalkoren reviewed (approved):
I do remember having the feeling that the actual merge reason is not fully followed through, but didn't find a proper test case. I hope there are no additional cases where hard-coded reason is used instead of the actual one.

Other than that, what a magnificent combination of code analysis, fixing and storytelling! I want to see the Gen AI application that does something like that 😉

andreidan committed:
Remove the merge method that received both the reason and merge context

This removes the (additional) `merge`  method with signature
```
public Mapper merge(Mapper mergeWith, MapperService.MergeReason reason, MapperMergeContext mapperMergeContext);
```
in favour of the existing
```
public abstract Mapper merge(Mapper mergeWith, MapperMergeContext mapperMergeContext);
```
as the `MapperMergeContext` arealdy contains the merge reason.

This will help the API clients avoid the pitfalls that already exist in tests where the provided
merge `reason` is not the same as the `MapperMergeContext#reason`

andreidan reviewed (commented):
Thanks for working on this Felix and for the awesome description (a thing of beauty!) 

I left some rather minor nits and a proposal to remove one of the `merge` methods which I sketched in a branch (to make sure it's not a humongous change - and it doesn't seem like it would be `+27, -61`): https://github.com/elastic/elasticsearch/commit/cdb4711517dacce6c4c5922ed16a02a833768300 

The purpose is to (hopefully) simplify the matrix of available public methods and avoid a potentially subtle bug where a `merge` method would be called with a `reason` that's not the same as the one configured in the mapper merge context 
We have already a few situations like this in tests (which I fixed in the [branch](https://github.com/elastic/elasticsearch/commit/cdb4711517dacce6c4c5922ed16a02a833768300) I sketched out the proposal in): 
```
        NestedObjectMapper result = (NestedObjectMapper) firstMapper.merge(
            secondMapper,
            MapperService.MergeReason.INDEX_TEMPLATE,
            MapperMergeContext.root(false, false, Long.MAX_VALUE)
        );
```

What do you think? (if you agree feel free to cherry-pick the commit or reimplement, as you wish - it's on a branch based on your branch so it should be straightforward to pick)

felixbarny committed:
Merge remote-tracking branch 'origin/main' into fix-template-merging-regression

felixbarny committed:
Add more tests

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperBuilderContext.java:
```diff
@@ -22,32 +23,39 @@ public class MapperBuilderContext {
      * The root context, to be used when building a tree of mappers
      */
     public static MapperBuilderContext root(boolean isSourceSynthetic, boolean isDataStream) {
-        return new MapperBuilderContext(null, isSourceSynthetic, isDataStream, false, ObjectMapper.Defaults.DYNAMIC);
+        return root(isSourceSynthetic, isDataStream, MergeReason.MAPPING_UPDATE);
+    }
+
+    public static MapperBuilderContext root(boolean isSourceSynthetic, boolean isDataStream, MergeReason mergeReason) {
+        return new MapperBuilderContext(null, isSourceSynthetic, isDataStream, false, ObjectMapper.Defaults.DYNAMIC, mergeReason);
```
Added in c0c5333f5ffcc7be6816560533f03de4ec0b2697

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -23,11 +25,18 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext, NewFieldsB
      */
-    public static MapperMergeContext root(boolean isSourceSynthetic, boolean isDataStream, long newFieldsBudget) {
-        return new MapperMergeContext(MapperBuilderContext.root(isSourceSynthetic, isDataStream), NewFieldsBudget.of(newFieldsBudget));
+    public static MapperMergeContext root(boolean isSourceSynthetic, boolean isDataStream, MergeReason mergeReason, long newFieldsBudget) {
```
Added in c0c5333f5ffcc7be6816560533f03de4ec0b2697.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -456,7 +456,7 @@ public final boolean subobjects() {
     @Override
     public ObjectMapper merge(Mapper mergeWith, MapperMergeContext mapperMergeContext) {
-        return merge(mergeWith, MergeReason.MAPPING_UPDATE, mapperMergeContext);
+        return merge(mergeWith, mapperMergeContext.getMapperBuilderContext().getMergeReason(), mapperMergeContext);
```
There's `ObjectMapperTests#testMerge` which already chooses a random merge reason. There are also three tests for the `INDEX_TEMPLATE` merge reason (`testMergeEnabledForIndexTemplates`, `testFieldReplacementForIndexTemplates`, `testDisallowFieldReplacementForIndexTemplates`). Does that cover what you had in mind?

felixbarny reviewed (commented):
> I left some rather minor nits and a proposal to remove one of the merge methods which I sketched in a branch (to make sure it's not a humongous change - and it doesn't seem like it would be +27, -61): https://github.com/elastic/elasticsearch/commit/cdb4711517dacce6c4c5922ed16a02a833768300

Thanks, I've cherry-picked your commit.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into fix-template-merging-regression

felixbarny committed:
Fix PassThroughObjectMapper by correctly overriding the merge method

andreidan reviewed (approved):
Thanks for iterating on this Felix. 

LGTM (once CI is happy) 🚀

felixbarny committed:
Rename test class to comply with convention

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.13  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 106077`

felixbarny commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.13|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/107254">](https://github.com/elastic/elasticsearch/pull/107254)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.4.6 -->
